training/COMPARISON.md

```diff
@@ -0,0 +1,462 @@
+
+#### GroupBy in NumPy
+While the concept is the same as Arkouda's `GroupBy`, `NumPy` does not explicitly support `GroupBy`. Instead, it uses `np.unique`.
+```python
+import numpy as np
+arr = np.array([4, 2, 5, 6, 4, 7, 2])
+keys, cts = np.unique(arr, return_counts=True)
+keys  # this represents the unique keys in the array
+array([2, 4, 5, 6, 7])
+cts  # this represents the number of times each key occurs in the array
+array([2, 2, 1, 1, 1])
+```
+
+Arkouda's `GroupBy` is a bit more advanced. Arkouda has a `GroupBy` class that leverages functionality similar to `np.unique` to compute its components. This class also includes `.permutaion` and `.segement` properties. `.permutation` is contains a pdarray whose values are the indexes of the original array in order by group. `.segments` contains a pdarray whose values are the starting index of each group (when `GroupBy.keys` is indexed by `GroupBy.permutation`).
+```python
+import arkouda as ak
+ak.connect()
+arr = ak.array([4, 2, 5, 6, 4, 7, 2])
+g = ak.GroupBy(arr)
+keys, cts = g.count()
+
+keys  # this represents the unique keys in the array
+array([2 4 5 6 7])
+cts  # this represents the number of times each key occurs in the array
+array([2 2 1 1 1])
+
+# the unique keys can also be accessed via the .unique_keys property
+g.unique_keys
+array([2 4 5 6 7])
+
+g.permutation
+array([1 6 0 4 2 3 5])
+g.segments
+array([0 2 4 5 6])
+g.keys[g.permutation]
+array([2 2 4 4 5 6 7])
+```
+
+Arkouda is able to create `GroupBy` objects from a sequence of groupable objects. 
+```python
+import arkouda as ak
+ak.connect()
+a = ak.array([0, 1, 2, 3, 3, 3, 2, 1])
+b = ak.array([0, 1, 1, 2, 3, 3, 1, 1])
+g = ak.GroupBy([a, b])
+keys, cts = g.count()
+
+keys  # this represents the unique keys in the array
+(array([0 1 2 3 3]), array([0 1 1 2 3]))  
+
+cts  # this represents the number of times each key occurs in the array
+array([1 2 2 1 2])
+
+g.permutation
+array([0 1 7 2 6 3 4 5])
+g.segments
+array([0 1 3 5 6])
+g.keys[0][g.permutation]
+array([0 1 1 2 2 3 3 3])
+g.keys[1][g.permutation]
+array([0 1 1 1 1 2 3 3])
+```
+Notice that keys is a tuple of length equal to the number of elements grouped on. The corresponding indexes are the keys. For Example, key 0 is `(0, 0)`.
+
+### Not Supported
+A common thing to do in Python is iterate over an object. This is not something that should be done (and is not directly supported) with Arkouda arrays.
+
+For example, attempting to interate directly on a pdarray will raise and exception.
+```python
+import arkouda as ak
+ak.connect()
+array = ak.arange(100)
+for i in array:
+    print(i)
+
+NotImplementedError: pdarray does not support iteration. To force data transfer from server, use to_ndarray
+```
+
+Unfortunately, there is a way to iterate `pdarray` objects without raising an exception. This is for informational purposes only. If you find yourself writing code similar to what appears below or wanting to iterate a `pdarray`, you will most likely need Chapel.
```

Comment:
there's likely a way to get the desired behavior with an array oriented method in arkouda or the functionality will need to be implemented in Chapel

Reply:
updated